src/inmanta/data/__init__.py

```diff
@@ -1620,6 +1620,29 @@ async def lock_table(cls, mode: TableLockMode, connection: asyncpg.connection.Co
         """
         await cls._execute_query(f"LOCK TABLE {cls.table_name()} IN {mode.value} MODE", connection=connection)
 
+    async def _xact_lock(
+        self, lock_key: int, instance_key: int, *, shared: bool = False, connection: asyncpg.Connection
+    ) -> None:
+        """
+        Acquires a transaction-level advisory lock for concurrency control
+
+        :param lock_key: the key identifying this lock (32 bit signed int)
+        :param instance_key: the key identifying the instance to lock (32 bit signed int)
+
+        :param shared: If true, doesn't conflict with other shared locks, only with non-shared ones.
+        :param connection: The connection hosting the transaction for which to acquire a lock.
+        """
+        lock: str = "pg_advisory_xact_lock_shared" if shared else "pg_advisory_xact_lock"
+        await connection.execute(
+            # Advisory lock keys are only 32 bit (or a single 64 bit key), while a full uuid is 128 bit.
+            # Since locking slightly too strictly at extremely low odds is acceptable, we only use a 32 bit subvalue
+            # of the uuid. For uuid4, time_low is (despite the name) randomly generated. Since it is an unsigned
+            # integer while Postgres expects a signed one, we shift it by 2**31.
```

Comment:
This whole block seized to be relevant in this context now that we don't pass an environment id anymore. Since this is a big part of the complexity, would it not be better to make it a method on environment as well instead of accepting the instance key, and mention in the docstring that the acquired lock may be in rare cases too strict with respect to the environment?

Reply:
tried to improve